### DISCUSSION TITLE
Specify destination root directory

### DISCUSSION THREAD

**Discussion by atlib77** [abraunegg/onedrive#1540] [Q&A] [answered]
Hello,

Is it possible to specify the destination root directory?

E.g. I want to save my home directory to onedrive under a subfolder, such as "laptop1".

That is:   $HOME  -->  <ONEDRIVE>/laptop1/

thanks!

**Comment by abraunegg** [comment, marked as answer]
@atlib77 
You will need to strategise your setup, and use symbolic links to achieve what you are trying to do. For example:

* 'sync_dir' will need to be configured to point to '/opt/OneDrive'
* Your user home directory symbolically linked to '/opt/OneDrive/laptop1/'

Refer to https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#upload-data-from-the-local-onedrive-folder-to-a-specific-location-on-onedrive for further details

Other factors:
* Use `--dry-run` to validate your local configuration
* Use `--local-first` to ensure your local folders are the source of truth

**Comment by atlib77** [comment]
Ah, that's nifty and should do the trick!  Thanks for the quick response!

**Comment by kokoko3k** [comment]
I was doing the following:
Say i want to sync my local:
/home/koko/dir1
/mnt/disc2/dir2
to:
onedrive:/subdir1/

What i would do is:
make local symlinks:
/home/koko/OneDrive/subdir1
link /home/koko/dir1 -> /home/koko/OneDrive/subdir1/dir1
link /mnt/disc2/dir2 -> /home/koko/OneDrive/subdir1/dir2

**But then i stopped and thought... wait, i've to leave symlinks enabled for this to work.
What if there are circular references? What if i don't want underlying symlinks to be uploaded?**

Is there another way or at least could you reconsider to upload to a specific folder instead, or maybe implement another simpler option to skip symlinks deeper than 1, maybe?

**Reply by abraunegg** [reply]
@kokoko3k 
> What if there are circular references? What if i don't want underlying symlinks to be uploaded?

OneDrive itself does not support symbolic links - it has no concept of them. A full file or folder must be uploaded. Thus, 'symbolic links' on a Unix|Linux|BSD platform need to be converted to an actual content.

It is not the responsibility of this client to sort out any circular references that you may have of that you have created yourself.

> Is there another way or at least could you reconsider to upload to a specific folder instead

The OneDrive API does not support this - thus the client cannot support this.

> or maybe implement another simpler option to skip symlinks deeper than 1, maybe

This is not possible - as - how to do you configure the application to skip the symbolic link in one location / but not another / but skip here / and not there .... 

You need to potentially rethink your file structure | folder structure locally if you are running into such issues.

**Reply by kokoko3k** [reply]
If I understood right, the use case is to let users to select where to put files remotely, right?
The suggested solution has been: "populate a folder with symlinks and upload that", which is fine, but requires you to not use the "--skip-symlinks", which i imagine is there for reasons.
So, if there is no intention of allowing an arbitrary directory to copy/sync the files to wia a specific option, but the use of symlinks is the only way, then, i propose a --skip-symlinks-after-depth option, so that one can still organize the folder to upload via symlinks till a certain level, and skip the deeper ones.

(Of course one can use bind mounts too, now that i think of it)

Anyway, I hope to have been clearer; what do you think?

**Reply by abraunegg** [reply]
@kokoko3k 
> So, if there is no intention of allowing an arbitrary directory to copy/sync the files to wia a specific option

Please understand it is not that I have no intention - the Microsoft provided API does not allow this. This is *not* my limitation.

> i propose a --skip-symlinks-after-depth option, so that one can still organize the folder to upload via symlinks till a certain level, and skip the deeper ones

Please submit a feature request filling in all the required details

**Comment by kokoko3k** [comment]
Well, somehow it has to be possible, since rclone is able to do it and i used it, so I can confirm it works, maybe it does some tricks under the hood:
```rclone sync $source $dest```

Btw I've made a feature request: https://github.com/abraunegg/onedrive/issues/1881

**Reply by abraunegg** [reply]
@kokoko3k 

> Well, somehow it has to be possible, since rclone is able to do it and i used it, so I can confirm it works, maybe it does some tricks under the hood: `rclone sync $source $dest`

rclone is a 1-way client - like an FTP client. This is why it is possible as it does not care about the entire state. This client is a 'sync' client keeping OneDrive in-sync with your local folder - and means keeping the data in the same spots. Do not confuse yourself that just because one client works like that - they all should/do.

**Reply by kokoko3k** [reply]
Absolutely, I never said this software should do it.

I don't know anything about onedrive protocol nor how this software works.
I was just answering to the statement: "the Microsoft provided API does not allow this." with mine: "(rclone) does some tricks under the hood".
One of the possible tricks is to use symlinks to "fake" a new directory structure; maybe rclone does the same in a way that is transparent to user (?)
Btw, I'm perfectly fine with the provided solution by you and I thank you for explaining better.
